File: app/lib/services/meal_memory/meal_reference_rules.dart
// Dart port of docs/meal_memory_starter/meal_reference_rules.py
//
// Domain rules for the pattern engine. All RegExp objects are compiled here
// (at class/const construction time) — not inside isReferential() calls.
// Add or tune rules here without touching reference_engine.dart.

import 'reference_engine.dart';

// ─── Temporal rules ───────────────────────────────────────────────────────────

final _temporalRules = <ReferenceRule>[
  ReferenceRule(
    key: 'two_days_ago',
    label: 'Two days ago reference',
    patterns: [
      RegExp(r'\bday before (?:last|yesterday)\b', caseSensitive: false),
      RegExp(r'\bnight before last\b', caseSensitive: false),
    ],
  ),
  ReferenceRule(
    key: 'yesterday',
    label: 'Yesterday reference',
    patterns: [
      RegExp(r'\byesterday\b', caseSensitive: false),
      RegExp(r'\blast night\b', caseSensitive: false),
      RegExp(r'\bthe (?:night|day) before\b', caseSensitive: false),
      RegExp(r'\bthe other night\b', caseSensitive: false),
    ],
  ),
  ReferenceRule(
    key: 'this_morning',
    label: 'This morning / earlier today',
    patterns: [
      RegExp(r'\bthis morning\b', caseSensitive: false),
      RegExp(r'\bearlier today\b', caseSensitive: false),
      RegExp(r'\bearlier\b', caseSensitive: false),
      RegExp(r'\ba few hours ago\b', caseSensitive: false),
    ],
  ),
  ReferenceRule(
    key: 'leftovers',
    label: 'Leftover reference',
    patterns: [
      RegExp(r'\bleftovers?\b', caseSensitive: false),
      RegExp(r'\bthe rest of\b', caseSensitive: false),
      RegExp(r"\bwhat(?:'s| was) left\b", caseSensitive: false),
    ],
  ),
  ReferenceRule(
    key: 'same_as_before',
    label: 'Same as a prior entry',
    patterns: [
      RegExp(r'\bsame [^.]{0,30}as\b', caseSensitive: false),
      RegExp(r'\bsame (?:thing|old|one)\b', caseSensitive: false),
      RegExp(r'\bagain\b', caseSensitive: false),
      RegExp(r'\brepeat\b', caseSensitive: false),
      RegExp(r'\b(?:the |my |as )?usual\b', caseSensitive: false),
      RegExp(r'\bmy go.?to\b', caseSensitive: false),
      RegExp(r'\bwhat i (?:always|usually|normally) (?:have|eat|get)\b', caseSensitive: false),
      RegExp(r'\bthe (?:thing|one) i (?:had|ate|got)\b', caseSensitive: false),
      RegExp(r'\blike (?:what |what I |I )?had\b', caseSensitive: false),
      RegExp(r'\bwhat i (?:had|ate)\b', caseSensitive: false),
    ],
  ),
  ReferenceRule(
    key: 'days_ago',
    label: 'N days ago',
    patterns: [
      RegExp(r'\b(?:a )?(?:few|couple(?:\s+of)?) (?:days|nights) ago\b', caseSensitive: false),
      RegExp(r'\b(?:a )?(?:few|couple(?:\s+of)?) days back\b', caseSensitive: false),
      RegExp(r'\b(?:two|three|four|five|2|3|4|5) (?:days|nights) ago\b', caseSensitive: false),
      RegExp(r'\bearlier (?:this|in the) week\b', caseSensitive: false),
      RegExp(r'\ba while (?:back|ago)\b', caseSensitive: false),
      RegExp(r'\blast week\b', caseSensitive: false),
    ],
  ),
  ReferenceRule(
    key: 'named_day',
    label: 'Named weekday reference',
    patterns: [
      RegExp(
        r'\b(?:on\s+)?(?:monday|tuesday|wednesday|thursday|friday|saturday|sunday)\b',
        caseSensitive: false,
      ),
      RegExp(r'\bthe other day\b', caseSensitive: false),
    ],
  ),
];

// ─── Meal type rules ──────────────────────────────────────────────────────────

final _mealTypeRules = <ReferenceRule>[
  ReferenceRule(
    key: 'meal_breakfast',
    label: 'Breakfast slot',
    patterns: [
      RegExp(r'\bbreakfast\b', caseSensitive: false),
      RegExp(r'\bmorning meal\b', caseSensitive: false),
      RegExp(r'\bbrunch\b', caseSensitive: false),
    ],
  ),
  ReferenceRule(
    key: 'meal_lunch',
    label: 'Lunch slot',
    patterns: [
      RegExp(r'\blunch\b', caseSensitive: false),
      RegExp(r'\bmidday\b', caseSensitive: false),
      RegExp(r'\bnoon\b', caseSensitive: false),
    ],
  ),
  ReferenceRule(
    key: 'meal_dinner',
    label: 'Dinner slot',
    patterns: [
      RegExp(r'\bdinner\b', caseSensitive: false),
      RegExp(r'\bsupper\b', caseSensitive: false),
      RegExp(r'\bevening meal\b', caseSensitive: false),
    ],
  ),
  ReferenceRule(
    key: 'meal_snack',
    label: 'Snack',
    patterns: [
      RegExp(r'\bsnack\b', caseSensitive: false),
      RegExp(r'\ba bite\b', caseSensitive: false),
      RegExp(r'\ba little something\b', caseSensitive: false),
    ],
  ),
];

// ─── Combined rule set + key sets ─────────────────────────────────────────────

final mealRules = [..._temporalRules, ..._mealTypeRules];

final temporalKeys = <String>{for (final r in _temporalRules) r.key};
final mealTypeKeys = <String>{for (final r in _mealTypeRules) r.key};

// ─── Query spec ───────────────────────────────────────────────────────────────

class MealQuerySpec {
  final int? dateOffset; // 0 = today, 1 = yesterday, 3 = ~3 days ago, null = unknown
  final String? mealType; // "breakfast" | "lunch" | "dinner" | "snack" | null
  final bool isLeftover;
  final bool matchRecent; // true → just get the N most recent entries

  const MealQuerySpec({
    this.dateOffset,
    this.mealType,
    this.isLeftover = false,
    this.matchRecent = false,
  });
}

MealQuerySpec buildQuerySpec(ReferenceProfile profile, {DateTime? now}) {
  final fired = profile.firedKeys.toSet();
  final effectiveNow = now ?? DateTime.now();

  int? dateOffset;
  if (fired.contains('two_days_ago')) {
    // "day before yesterday" fires both two_days_ago and yesterday — two_days_ago wins
    dateOffset = 2;
  } else if (fired.contains('yesterday') || fired.contains('leftovers')) {
    dateOffset = 1;
  } else if (fired.contains('days_ago')) {
    // days_ago before this_morning: "earlier this week" fires both, days_ago wins
    dateOffset = 3;
  } else if (fired.contains('this_morning')) {
    dateOffset = 0;
  }

  // named_day overrides dateOffset — "leftovers from last friday" means friday,
  // not yesterday. Walk backwards from yesterday to find the closest matching day.
  if (fired.contains('named_day')) {
    final resolved = _resolveNamedDayOffset(profile.inputText, effectiveNow);
    if (resolved != null) dateOffset = resolved;
  }

  final matchRecent = dateOffset == null &&
      (fired.contains('same_as_before') || fired.contains('named_day'));

  String? mealType;
  for (final pair in [
    ('meal_breakfast', 'breakfast'),
    ('meal_lunch', 'lunch'),
    ('meal_dinner', 'dinner'),
    ('meal_snack', 'snack'),
  ]) {
    if (fired.contains(pair.$1)) {
      mealType = pair.$2;
      break;
    }
  }

  return MealQuerySpec(
    dateOffset: dateOffset,
    mealType: mealType,
    isLeftover: fired.contains('leftovers') || fired.contains('same_as_before'),
    matchRecent: matchRecent,
  );
}

// Walk backwards from yesterday, find the most recent occurrence of the named
// weekday. Returns days ago (1–14 max). No modular math — just proximity.
int? _resolveNamedDayOffset(String normalizedInput, DateTime now) {
  const days = [
    ('monday', DateTime.monday),
    ('tuesday', DateTime.tuesday),
    ('wednesday', DateTime.wednesday),
    ('thursday', DateTime.thursday),
    ('friday', DateTime.friday),
    ('saturday', DateTime.saturday),
    ('sunday', DateTime.sunday),
  ];

  int? targetWeekday;
  for (final (name, weekday) in days) {
    if (normalizedInput.contains(name)) {
      targetWeekday = weekday;
      break;
    }
  }
  if (targetWeekday == null) return null;

  for (var i = 1; i <= 14; i++) {
    final candidate = now.subtract(Duration(days: i));
    if (candidate.weekday == targetWeekday) return i;
  }
  return null;
}

File: docs/meal_memory_starter/meal_reference_rules.py
"""Meal memory reference rules — domain layer for pattern_engine.py.

Each ReferenceRule defines a key, a human label, and a tuple of regex patterns
that indicate the user is referencing a past meal. Add, remove, or tune rules
here without touching pattern_engine.py.

Rule keys are grouped into two sets:
  TEMPORAL_KEYS  — signals that the user is referring to a past point in time
  MEAL_TYPE_KEYS — signals which meal slot (breakfast/lunch/dinner/snack) they mean

detect_references() uses these sets to set has_temporal_ref and has_meal_type
on the returned ReferenceProfile, which drives what DB query to build.

Pattern guidelines:
  - \b word boundaries prevent "yesterday" matching "the day before yesterday" twice
  - [^.]{0,80} limits cross-sentence false positives
  - Keep patterns specific enough to avoid matching neutral food descriptions
  - Add new rules for slang your users actually say ("the usual", "my go-to")
"""

from __future__ import annotations

from pattern_engine import ReferenceRule, compile_patterns


# ---------------------------------------------------------------------------
# Temporal reference rules
# Fired when the user refers to a past point in time.
# ---------------------------------------------------------------------------

_TEMPORAL_RULES: tuple[ReferenceRule, ...] = (
    ReferenceRule(
        key="yesterday",
        label="Yesterday reference",
        patterns=compile_patterns((
            r"\byesterday\b",
            r"\blast night\b",
            r"\bthe night before\b",
        )),
    ),
    ReferenceRule(
        key="this_morning",
        label="This morning / earlier today",
        patterns=compile_patterns((
            r"\bthis morning\b",
            r"\bearlier today\b",
            r"\bearlier\b",
            r"\ba few hours ago\b",
        )),
    ),
    ReferenceRule(
        key="leftovers",
        label="Leftover reference",
        patterns=compile_patterns((
            r"\bleftovers?\b",
            r"\bthe rest of\b",
            r"\bwhat(?:'s| was) left\b",
        )),
    ),
    ReferenceRule(
        key="same_as_before",
        label="Same as a prior entry",
        patterns=compile_patterns((
            r"\bsame [^.]{0,30}as\b",  # "same X as", "same thing as"
            r"\bsame thing\b",
            r"\bagain\b",
            r"\b(?:the |my |as )?usual\b",  # "the usual", "my usual", "as usual", "usual"
            r"\bmy go.?to\b",
            r"\bwhat i (?:always|usually|normally) (?:have|eat|get)\b",
        )),
    ),
    ReferenceRule(
        key="days_ago",
        label="N days ago",
        patterns=compile_patterns((
            r"\b(?:a )?(?:few|couple(?:\s+of)?) days ago\b",
            r"\b(?:two|three|four|five|2|3|4|5) days ago\b",
            r"\bearlier this week\b",
        )),
    ),
    ReferenceRule(
        key="named_day",
        label="Named weekday reference",
        patterns=compile_patterns((
            r"\b(?:on\s+)?(?:monday|tuesday|wednesday|thursday|friday|saturday|sunday)\b",
            r"\bthe other day\b",
        )),
    ),
)

# ---------------------------------------------------------------------------
# Meal type rules
# Fired when the user names a meal slot — helps narrow the DB query.
# ---------------------------------------------------------------------------

_MEAL_TYPE_RULES: tuple[ReferenceRule, ...] = (
    ReferenceRule(
        key="meal_breakfast",
        label="Breakfast slot",
        patterns=compile_patterns((
            r"\bbreakfast\b",
            r"\bmorning meal\b",
            r"\bbrunch\b",
        )),
    ),
    ReferenceRule(
        key="meal_lunch",
        label="Lunch slot",
        patterns=compile_patterns((
            r"\blunch\b",
            r"\bmidday\b",
            r"\bnoon\b",
        )),
    ),
    ReferenceRule(
        key="meal_dinner",
        label="Dinner slot",
        patterns=compile_patterns((
            r"\bdinner\b",
            r"\bsupper\b",
            r"\bevening meal\b",
        )),
    ),
    ReferenceRule(
        key="meal_snack",
        label="Snack",
        patterns=compile_patterns((
            r"\bsnack\b",
            r"\ba bite\b",
            r"\ba little something\b",
        )),
    ),
)

# ---------------------------------------------------------------------------
# Combined rule set and key sets
# ---------------------------------------------------------------------------

RULES: tuple[ReferenceRule, ...] = _TEMPORAL_RULES + _MEAL_TYPE_RULES

TEMPORAL_KEYS: frozenset[str] = frozenset(r.key for r in _TEMPORAL_RULES)
MEAL_TYPE_KEYS: frozenset[str] = frozenset(r.key for r in _MEAL_TYPE_RULES)


# ---------------------------------------------------------------------------
# Convenience: build a DB query spec from a ReferenceProfile
# ---------------------------------------------------------------------------

from pattern_engine import ReferenceProfile  # noqa: E402 — post-class import ok


def build_query_spec(profile: ReferenceProfile) -> dict:
    """Translate a ReferenceProfile into a structured DB query spec.

    Returns a dict the caller can use to query the meal_fingerprints table.
    The caller decides how to execute the query (SQLite, Drift, etc.).

    Shape:
        {
          "date_offset": int | None,   # 0 = today, 1 = yesterday, None = unknown
          "meal_type": str | None,     # "breakfast" | "lunch" | "dinner" | "snack" | None
          "is_leftover": bool,         # True → same meal, different entry
          "match_recent": bool,        # True → just get the N most recent entries
        }
    """
    spec: dict = {
        "date_offset": None,
        "meal_type": None,
        "is_leftover": False,
        "match_recent": False,
    }

    fired = set(profile.fired_keys)

    # Date offset
    if "yesterday" in fired or "leftovers" in fired:
        spec["date_offset"] = 1          # yesterday
    elif "this_morning" in fired:
        spec["date_offset"] = 0          # today (earlier slot)
    elif "days_ago" in fired:
        spec["date_offset"] = 3          # rough midpoint; caller can refine
    elif "same_as_before" in fired or "named_day" in fired:
        spec["match_recent"] = True

    # Meal type
    for key, meal_type in (
        ("meal_breakfast", "breakfast"),
        ("meal_lunch", "lunch"),
        ("meal_dinner", "dinner"),
        ("meal_snack", "snack"),
    ):
        if key in fired:
            spec["meal_type"] = meal_type
            break

    # Leftover flag
    if "leftovers" in fired or "same_as_before" in fired:
        spec["is_leftover"] = True

    return spec
